File: scripts/sparge_home.py
"""
Sparge home directory management.

Reads ~/.sparge/config.json to find where project data lives.
Creates the config with defaults on first use.

Public API
----------
get_projects_dir() -> Path   resolved path to the projects directory
"""
from __future__ import annotations

import json
import os
from pathlib import Path


_SPARGE_HOME = Path.home() / '.sparge'
_SPARGE_CFG  = _SPARGE_HOME / 'config.json'
_DEFAULT_PROJECTS_DIR = Path.home() / 'sparge-projects'


def get_projects_dir() -> Path:
    """
    Return the resolved projects directory path.
    Creates ~/.sparge/config.json with defaults if absent.
    """
    _SPARGE_HOME.mkdir(exist_ok=True)
    if not _SPARGE_CFG.exists():
        _SPARGE_CFG.write_text(json.dumps(
            {'projects_dir': str(_DEFAULT_PROJECTS_DIR)},
            indent=2,
        ))
    try:
        data = json.loads(_SPARGE_CFG.read_text())
        raw = data.get('projects_dir', str(_DEFAULT_PROJECTS_DIR))
    except Exception:
        raw = str(_DEFAULT_PROJECTS_DIR)
    return Path(os.path.expanduser(raw))

File: server/src/main/java/io/sparge/server/SpargeHome.java
package io.sparge.server;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import jakarta.enterprise.context.ApplicationScoped;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Manages ~/.sparge/config.json — machine-wide Sparge home config.
 * Mirrors scripts/sparge_home.py.
 *
 * Accepts an injected home path for testability; CDI usage passes the real home dir.
 */
@ApplicationScoped
public class SpargeHome {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final Path home;
    private final Path spargeDir;
    private final Path spargeCfg;
    private final Path defaultProjectsDir;

    /** CDI no-arg constructor — uses real user home. */
    public SpargeHome() {
        this(Path.of(System.getProperty("user.home")));
    }

    /** Testable constructor — accepts any home dir. */
    SpargeHome(Path home) {
        this.home               = home;
        this.spargeDir          = home.resolve(".sparge");
        this.spargeCfg          = spargeDir.resolve("config.json");
        this.defaultProjectsDir = home.resolve("sparge-projects");
    }

    /**
     * Return the resolved projects directory.
     * Creates ~/.sparge/config.json with defaults if absent.
     */
    public Path getProjectsDir() throws IOException {
        Files.createDirectories(spargeDir);
        if (!Files.exists(spargeCfg)) {
            ObjectNode cfg = MAPPER.createObjectNode();
            cfg.put("projects_dir", defaultProjectsDir.toString());
            MAPPER.writerWithDefaultPrettyPrinter().writeValue(spargeCfg.toFile(), cfg);
        }
        try {
            ObjectNode data = (ObjectNode) MAPPER.readTree(spargeCfg.toFile());
            String raw = data.has("projects_dir")
                    ? data.get("projects_dir").asText()
                    : defaultProjectsDir.toString();
            return Path.of(raw.replace("~", home.toString()));
        } catch (Exception e) {
            return defaultProjectsDir;
        }
    }
}
